components/server/src/api/workspace-service-api.ts

```diff
@@ -92,4 +99,80 @@ export class WorkspaceServiceAPI implements ServiceImpl<typeof WorkspaceServiceI
         response.pagination.total = resultTotal;
         return response;
     }
+
+    async createAndStartWorkspace(req: CreateAndStartWorkspaceRequest): Promise<CreateAndStartWorkspaceResponse> {
+        // We rely on FGA to do the permission checking
+        if (req.source?.case !== "contextUrl") {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "not implemented");
+        }
+        if (!req.organizationId || !uuidValidate(req.organizationId)) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "organizationId is required");
+        }
+        if (!req.editor) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "editor is required");
+        }
+        if (!req.source.value) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "value is required in contextUrl source");
+        }
+        const contextUrl = req.source.value;
+        const user = await this.userService.findUserById(ctxUserId(), ctxUserId());
+        const { context, project } = await this.contextService.parseContext(user, contextUrl, {
+            projectId: req.configurationId,
+            organizationId: req.organizationId,
+            forceDefaultConfig: req.forceDefaultConfig,
+        });
+
+        await this.workspaceService.mayStartWorkspace({}, user, req.organizationId, Promise.resolve([]));
+        const normalizedContextUrl = this.contextParser.normalizeContextURL(contextUrl);
+        const workspace = await this.workspaceService.createWorkspace(
+            {},
+            user,
+            req.organizationId,
+            project,
+            context,
+            normalizedContextUrl,
+        );
+
+        await this.workspaceService.startWorkspace({}, user, workspace.id, {
+            forceDefaultImage: req.forceDefaultConfig,
+            workspaceClass: req.workspaceClass,
+            ideSettings: {
+                defaultIde: req.editor.name,
+                useLatestVersion: req.editor.version === "latest",
+            },
+            clientRegionCode: ctxClientRegion(),
+        });
+
+        const info = await this.workspaceService.getWorkspace(ctxUserId(), workspace.id);
+        const response = new CreateAndStartWorkspaceResponse();
+        response.workspace = this.apiConverter.toWorkspace(info);
+        return response;
+    }
+
+    async startWorkspace(req: StartWorkspaceRequest): Promise<StartWorkspaceResponse> {
+        // We rely on FGA to do the permission checking
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        const user = await this.userService.findUserById(ctxUserId(), ctxUserId());
+        const { workspace, latestInstance: instance } = await this.workspaceService.getWorkspace(
+            ctxUserId(),
+            req.workspaceId,
+        );
+        if (instance && instance.status.phase !== "stopped") {
```

Comment:
is it really an expectation that we return an existing workspace if it is already running?

it would be good to document